File: app/src/main/java/com/jchanghong/ActivityNoteEdit.java
package com.jchanghong;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.Snackbar;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.jchanghong.data.DatabaseManager;
import com.jchanghong.model.Category;
import com.jchanghong.model.Note;
import com.jchanghong.utils.Tools;

import com.jchanghong.R;

public class ActivityNoteEdit extends AppCompatActivity {

    public static final String EXTRA_OBJCT = "com.jchanghong.EXTRA_OBJECT_NOTE";
    public static final int OPEN_DIALOG_CATEGORY_CODE = 1;

    private Toolbar toolbar;
    private ActionBar actionBar;
    private View parent_view;
    private EditText tittle, content;
    private TextView time;
    private ImageView cat_icon, cat_drop;
    private TextView cat_text;
    private AppBarLayout appbar;
    private Menu menu;

    private boolean fav_checked = false;
    private boolean is_new = true;

    private Note ext_note = null;
    private Category cur_category = null;
    private DatabaseManager db;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_note);
        parent_view = findViewById(android.R.id.content);

        // init db
        db = new DatabaseManager(this);

        initToolbar();

        // get extra object
        ext_note = (Note) getIntent().getSerializableExtra(EXTRA_OBJCT);
        iniComponent();

    }


    private void initToolbar() {
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setHomeButtonEnabled(true);
        actionBar.setTitle("");
    }


    private void iniComponent() {
        tittle = (EditText) findViewById(R.id.tittle);
        content = (EditText) findViewById(R.id.content);
        time = (TextView) findViewById(R.id.time);

        cat_icon = (ImageView) findViewById(R.id.cat_icon);
        cat_drop = (ImageView) findViewById(R.id.cat_drop);
        cat_text = (TextView) findViewById(R.id.cat_text);
        appbar = (AppBarLayout) findViewById(R.id.appbar);

        is_new = (ext_note == null);

        if (is_new) {
            time.setText("");
            cur_category = db.getFirstCategory();
        } else {
            time.setText("Edited : " + Tools.stringToDate(ext_note.getLastEdit()));
            tittle.setText(ext_note.getTittle());
            content.setText(ext_note.getContent());
            cur_category = ext_note.getCategory();
        }
        setCategoryView(cur_category);

        ((LinearLayout) findViewById(R.id.lyt_category)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent i = new Intent(getApplicationContext(), ActivityCategoryPick.class);
                startActivityForResult(i, OPEN_DIALOG_CATEGORY_CODE);
            }
        });

        ((Button) findViewById(R.id.bt_save)).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                actionSave();
            }
        });
    }

    private void setCategoryView(Category category) {
        cat_icon.setImageResource(Tools.StringToResId(category.getIcon(),getApplicationContext()));
        cat_icon.setColorFilter(Color.parseColor(category.getColor()));
        cat_drop.setColorFilter(Color.parseColor(category.getColor()));
        cat_text.setText(category.getName());
        cat_text.setTextColor(Color.parseColor(category.getColor()));
        appbar.setBackgroundColor(Color.parseColor(category.getColor()));
        Tools.systemBarLolipopCustom(ActivityNoteEdit.this, category.getColor());
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        this.menu = menu;
        getMenuInflater().inflate(R.menu.menu_activity_manage_note, menu);
        //set fav icon
        if (!is_new) {
            if (ext_note.getFavourite() == 1) {
                fav_checked = true;
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ic_favorites_solid));
            } else {
                fav_checked = false;
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ic_favorites_outline));
            }
        } else {
            menu.getItem(0).setVisible(false);
            menu.getItem(1).setVisible(false);
        }
        return true;
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == OPEN_DIALOG_CATEGORY_CODE && resultCode == Activity.RESULT_OK) {
            cur_category = (Category) data.getSerializableExtra(ActivityCategoryPick.EXTRA_OBJ);
            setCategoryView(cur_category);
            Snackbar.make(parent_view, "Category Selected : " + cur_category.getName(), Snackbar.LENGTH_SHORT).show();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
            case R.id.action_delete:
                deleteConfirmation();
                break;
            case R.id.action_fav:
                actionFavorite();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void actionFavorite() {
        if (!is_new) {
            if (fav_checked) {
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ic_favorites_outline));
                db.removeFav(ext_note.getId());
                Snackbar.make(parent_view, "Removed from favorites", Snackbar.LENGTH_SHORT).show();
                fav_checked = false;
            } else {
                menu.getItem(0).setIcon(getResources().getDrawable(R.drawable.ic_favorites_solid));
                db.setFav(ext_note.getId());
                Snackbar.make(parent_view, "Added to favorites", Snackbar.LENGTH_SHORT).show();
                fav_checked = true;
            }
        }
    }

    private void actionSave() {
        if (tittle.getText().toString().equals("") || content.getText().toString().equals("")) {
            Snackbar.make(parent_view, "Tittle or Content can't be empty", Snackbar.LENGTH_SHORT).show();
        } else {
            if (is_new) ext_note = new Note();
            String notif_text;

            ext_note.setTittle(tittle.getText().toString());
            ext_note.setContent(content.getText().toString());
            ext_note.setLastEdit(System.currentTimeMillis());
            ext_note.setCategory(cur_category);

            if (is_new) {
                notif_text = "Note Saved";
                db.insertNote(ext_note);
            } else {
                notif_text = "Note Updated";
                db.updateNote(ext_note);
            }

            Snackbar.make(parent_view, notif_text, Snackbar.LENGTH_SHORT).setCallback(new Snackbar.Callback() {
                @Override
                public void onDismissed(Snackbar snackbar, int event) {
                    super.onDismissed(snackbar, event);
                    finish();
                }
            }).show();
        }
    }

    @Override
    public void onBackPressed() {
        if (ext_note != null) {
            if ((!(tittle.getText().toString().equals(ext_note.getTittle()))) ||
                    (!(content.getText().toString().equals(ext_note.getContent()))) ||
                    (!(cur_category.getId() == ext_note.getCategory().getId()))
                    ) {
                backConfirmation();
            } else {
                finish();
            }
        } else {
            if (tittle.getText().toString().equals("") && content.getText().toString().equals("")) {
                //do nothing
                finish();
            } else {
                backConfirmation();
            }
        }
    }

    private void backConfirmation() {
        AlertDialog.Builder builder = new AlertDialog.Builder(ActivityNoteEdit.this);
        builder.setTitle("Save Confirmation");
        builder.setMessage("Do you want to save?");
        builder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                if (ext_note == null) {
                    //save new note
                    Note n = new Note();
                    n.setTittle(tittle.getText() + "");
                    n.setContent(content.getText() + "");
                    n.setCategory(cur_category);
                    n.setFavourite(0);
                    n.setLastEdit(System.currentTimeMillis());
                    db.insertNote(n);
                } else {
                    ext_note.setTittle(tittle.getText() + "");
                    ext_note.setContent(content.getText() + "");
                    ext_note.setCategory(cur_category);
                    //no need to set fav here, fav already save to DB when clicked
                    ext_note.setLastEdit(System.currentTimeMillis());
                    db.updateNote(ext_note);
                    ext_note.clear();
                }
                Snackbar.make(parent_view, "Note Saved", Snackbar.LENGTH_SHORT).show();
                finish();
            }
        });
        builder.setNegativeButton("No", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                finish();
            }
        });
        builder.show();
    }

    private void deleteConfirmation() {
        AlertDialog.Builder builder = new AlertDialog.Builder(ActivityNoteEdit.this);
        builder.setTitle("Delete Confirmation");
        builder.setMessage("Are you sure want to delete this Note?");
        builder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                if (ext_note != null) {//modify
                    db.deleteNote(ext_note.getId());
                }
                Toast.makeText(getApplicationContext(), "Note Deleted", Toast.LENGTH_SHORT).show();
                finish();
            }
        });
        builder.setNegativeButton("No", null);
        builder.show();
    }

}

File: app/src/main/java/com/jchanghong/ActivityNoteEdit.kt
package com.jchanghong

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.design.widget.Snackbar
import android.support.v7.app.ActionBar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import com.jchanghong.data.Constant
import com.jchanghong.data.DatabaseManager
import com.jchanghong.model.Category
import com.jchanghong.model.Note
import com.jchanghong.utils.Tools
import com.jchanghong.utils.getyam
import com.jchanghong.utils.hasYamHead
import com.jchanghong.utils.removeyam
import gr.tsagi.jekyllforandroid.app.activities.PreviewMarkdownActivity
import gr.tsagi.jekyllforandroid.app.utils.GithubPush
import org.yaml.snakeyaml.Yaml
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutionException

class ActivityNoteEdit : AppCompatActivity() {

  lateinit  private var toolbar: Toolbar
  lateinit  private var actionBar: ActionBar
  lateinit  private var parent_view: View
   lateinit private var tittle: EditText
  lateinit  private var content: EditText
  lateinit  private var time: TextView
  lateinit  private var cat_icon: ImageView
  lateinit  private var cat_drop: ImageView
  lateinit  private var cat_text: TextView
  lateinit  private var appbar: AppBarLayout
   lateinit private var menu: Menu

    private var fav_checked = false
    private var is_new = true

    private var ext_note: Note? = null
    private var cur_category: Category? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_note)
        parent_view = findViewById(android.R.id.content)

        initToolbar()
        // get extra object
        ext_note = intent.getSerializableExtra(EXTRA_OBJCT) as? Note
        iniComponent()

    }


    private fun initToolbar() {
        toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        actionBar = supportActionBar!!
        actionBar .setDisplayHomeAsUpEnabled(true)
        actionBar .setHomeButtonEnabled(true)
        actionBar .title = ""
    }


    private fun iniComponent() {
        tittle = findViewById(R.id.tittle) as EditText
        content = findViewById(R.id.content) as EditText
        time = findViewById(R.id.time) as TextView

        cat_icon = findViewById(R.id.cat_icon) as ImageView
        cat_drop = findViewById(R.id.cat_drop) as ImageView
        cat_text = findViewById(R.id.cat_text) as TextView
        appbar = findViewById(R.id.appbar) as AppBarLayout

        is_new = ext_note == null

        if (is_new) {
            time .text = ""
            cur_category = DatabaseManager .defaultCAT
        } else {
            time .text = getString(R.string.time_edited) + Tools.stringToDate(ext_note?.lastEdit?:1)
            tittle .setText(ext_note ?.tittle?:"")
            content .setText(ext_note ?.content?.removeyam()?:"null")
            cur_category = ext_note ?.category
        }
        setCategoryView(cur_category?:DatabaseManager.defaultCAT)

        (findViewById(R.id.lyt_category) as LinearLayout).setOnClickListener {
            val i = Intent(applicationContext, ActivityCategoryPick::class.java)
            startActivityForResult(i, OPEN_DIALOG_CATEGORY_CODE)
        }

        (findViewById(R.id.bt_save) as Button).setOnClickListener { actionSave() }
    }

    private fun setCategoryView(category: Category) {
        cat_icon .setImageResource(Tools.StringToResId(category.icon, applicationContext))
        cat_icon .setColorFilter(Color.parseColor(category.color))
        cat_drop .setColorFilter(Color.parseColor(category.color))
        cat_text .text = category.name
        cat_text .setTextColor(Color.parseColor(category.color))
        appbar .setBackgroundColor(Color.parseColor(category.color))
        Tools.systemBarLolipopCustom(this@ActivityNoteEdit, category.color)
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        this.menu = menu
        menuInflater.inflate(R.menu.menu_activity_manage_note, menu)
        //set fav icon
        if (!is_new) {
            if (ext_note ?.favourite == 1) {
                fav_checked = true
                menu.getItem(0).icon = resources.getDrawable(R.drawable.ic_favorites_solid,theme)
            } else {
                fav_checked = false
                menu.getItem(0).icon = resources.getDrawable(R.drawable.ic_favorites_outline,theme)
            }
        } else {
            menu.getItem(0).isVisible = false
            menu.getItem(1).isVisible = false
        }
        return true
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == OPEN_DIALOG_CATEGORY_CODE && resultCode == Activity.RESULT_OK) {
            cur_category = data?.getSerializableExtra(ActivityCategoryPick.EXTRA_OBJ) as? Category
            setCategoryView(cur_category?:DatabaseManager.firstCategory)
            if (Constant.iszhong) {

                Snackbar.make(parent_view, "选择了分类 : " + cur_category?.name, Snackbar.LENGTH_SHORT).show()
            }
            else{

                Snackbar.make(parent_view, "Category Selected : " + cur_category?.name, Snackbar.LENGTH_SHORT).show()
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            android.R.id.home -> onBackPressed()
            R.id.action_delete -> deleteConfirmation()
            R.id.action_fav -> actionFavorite()
            R.id.action_preview-> preview()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun preview() {
            val myIntent = Intent(this, PreviewMarkdownActivity::class.java)
            myIntent.putExtra(PreviewMarkdownActivity.POST_CONTENT,ext_note?.content?.removeyam())
            startActivity(myIntent)
    }

    private fun actionFavorite() {
        if (!is_new) {
            if (fav_checked) {
                menu .getItem(0).icon = resources.getDrawable(R.drawable.ic_favorites_outline,theme)
                DatabaseManager .removeFav(ext_note?.id?:1)
                Snackbar.make(parent_view, getString(R.string.Removedfromfavorites), Snackbar.LENGTH_SHORT).show()
                fav_checked = false
            } else {
                menu .getItem(0).icon = resources.getDrawable(R.drawable.ic_favorites_solid,theme)
                DatabaseManager .setFav(ext_note?.id?:1)
                Snackbar.make(parent_view, getString(R.string.Addedtofavorites), Snackbar.LENGTH_SHORT).show()
                fav_checked = true
            }
        }
    }

    private fun actionSave() {
        if (tittle .text.toString() == "" || content .text.toString() == "") {
            Snackbar.make(parent_view, getString(R.string.Contentcannotbeempty), Snackbar.LENGTH_SHORT).show()
        } else {
            if (is_new) ext_note = Note()
            val notif_text: String

            ext_note ?.tittle = tittle .text.toString()
            ext_note ?.content = content .text.toString()
            ext_note ?.lastEdit = System.currentTimeMillis()
            ext_note ?.category = cur_category!!

            if (is_new) {
                notif_text = getString(R.string.notesaved)
                DatabaseManager .insertNote(ext_note!!)
                pushPost(ext_note!!)
            } else {
                notif_text = getString(R.string.noteupdate)
                DatabaseManager .updateNote(ext_note!!)
                pushPost(ext_note!!)
            }

            Snackbar.make(parent_view, notif_text, Snackbar.LENGTH_SHORT).setCallback(object : Snackbar.Callback() {
                override fun onDismissed(snackbar: Snackbar?, event: Int) {
                    super.onDismissed(snackbar, event)
                    finish()
                }
            }).show()
        }
    }

    fun pushPost(note: Note) {

        val date = SimpleDateFormat("yyyy-MM-dd").format(Date(note.lastEdit))
        val pusher = GithubPush(this)
        if (note.content.hasYamHead()) {
            pusher.pushContent(note.tittle, date, note.content,note.category.name)
        }
        val yaml = Yaml()
//        val customYaml = prefs.getString("yaml_values", "")
//        Log.d("yaml", customYaml)
        var map = yaml.load(note.content.getyam()) as? HashMap<String, Any>
        if (map==null)
        {
            map = HashMap<String, Any>()
            map.put("tags", note.category.name)
        }
//        map.put("tags", note.category.name)
        map.put("category", note.category.name)
        map.put("title", note.tittle)
        map.put("layout", "post")
        val output = "---\n" + yaml.dump(map) + "---\n"

        try {
            pusher.pushContent(note.tittle, date, output + note.content.removeyam(),note.category.name)

        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }


    override fun onBackPressed() {
        if (ext_note != null) {
            if (tittle .text.toString() != ext_note!!.tittle ||
                    content .text.toString() != ext_note!!.content.removeyam() ||
                    cur_category?.id != ext_note!!.category.id) {
                backConfirmation()
            } else {
                finish()
            }
        } else {
            if (tittle .text.toString() == "" && content .text.toString() == "") {
                //do nothing
                finish()
            } else {
                backConfirmation()
            }
        }
    }

    private fun backConfirmation() {
        val builder = AlertDialog.Builder(this@ActivityNoteEdit)
        builder.setTitle(getString(R.string.SaveConfirmation))
        builder.setMessage(getString(R.string.doyouwanttosave))
        builder.setPositiveButton(getString(R.string.yes)) { dialogInterface, i ->
            if (ext_note == null) {
                //save new note
                val n = Note()
                n.tittle = tittle .text.toString() + ""
                n.content = content .text.toString() + ""
                n.category = cur_category!!
                n.favourite = 0
                n.lastEdit = System.currentTimeMillis()
                DatabaseManager .insertNote(n)
                pushPost(n)
            } else {
                ext_note?.tittle = tittle .text.toString() + ""
                ext_note?.content = content .text.toString() + ""
                ext_note?.category = cur_category!!
                //no need to set fav here, fav already save to DB when clicked
                ext_note?.lastEdit = System.currentTimeMillis()
                DatabaseManager .updateNote(ext_note!!)
                pushPost(ext_note!!)
                ext_note?.clear()
            }
            Snackbar.make(parent_view, getString(R.string.notesaved), Snackbar.LENGTH_SHORT).show()
            finish()
        }
        builder.setNegativeButton("No") { dialogInterface, i -> finish() }
        builder.show()
    }

    private fun deleteConfirmation() {
        val builder = AlertDialog.Builder(this@ActivityNoteEdit)
        builder.setTitle(getString(R.string.SaveConfirmation1))
        builder.setMessage(getString(R.string.areyouwantdelete))
        builder.setPositiveButton(getString(R.string.yes)) { dialogInterface, i ->
            if (ext_note != null) {//modify
                DatabaseManager .deleteNote(ext_note?.id?:1)
            }
            Toast.makeText(applicationContext, getString(R.string.notedeleted), Toast.LENGTH_SHORT).show()
            finish()
        }
        builder.setNegativeButton(getString(R.string.no), null)
        builder.show()
    }

    companion object {

        val EXTRA_OBJCT = "com.jchanghong.EXTRA_OBJECT_NOTE"
        val OPEN_DIALOG_CATEGORY_CODE = 1
    }

}
